### PR TITLE
Use ArrayBuilders instead of Lists in more places.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

File: src/Features/Core/Portable/AddImport/SymbolReferenceFinder.cs [hunk 1]
```diff
@@ -119,13 +119,11 @@ private async Task<ImmutableArray<SymbolReference>> DoAsync(SearchScope searchSc
...
 
                 // Spin off tasks to do all our searching in parallel
-                var tasks = new List<Task<ImmutableArray<SymbolReference>>>
-                {
-                    GetReferencesForMatchingTypesAsync(searchScope),
-                    GetReferencesForMatchingNamespacesAsync(searchScope),
-                    GetReferencesForMatchingFieldsAndPropertiesAsync(searchScope),
-                    GetReferencesForMatchingExtensionMethodsAsync(searchScope),
-                };
+                using var _1 = ArrayBuilder<Task<ImmutableArray<SymbolReference>>>.GetInstance(out var tasks);
```
Comment (sharwell) [inline src/Features/Core/Portable/AddImport/SymbolReferenceFinder.cs:122]: 📝 this one can just be a fixed-size local array allocation. The length is either 4 or 8 depending on `searchScope.Exact`.

File: src/Features/Core/Portable/AddParameter/AddParameterService.cs [hunk 2]
```diff
@@ -269,11 +270,11 @@ public bool HasCascadingDeclarations(IMethodSymbol method)
             }
         }
 
-        private static List<SyntaxTrivia> GetDesiredLeadingIndentation(
+        private static ImmutableArray<SyntaxTrivia> GetDesiredLeadingIndentation(
             SyntaxGenerator generator, ISyntaxFactsService syntaxFacts,
             SyntaxNode node, bool includeLeadingNewLine)
         {
-            var triviaList = new List<SyntaxTrivia>();
+            using var _ = ArrayBuilder<SyntaxTrivia>.GetInstance(out var triviaList);
```
Comment (sharwell) [inline src/Features/Core/Portable/AddParameter/AddParameterService.cs:277]: 📝 This builder is not actually needed. The result will either be a 1- or 2-element array.

File: src/Features/Core/Portable/CodeRefactorings/CodeRefactoringService.cs [hunk 3]
```diff
@@ -125,7 +125,7 @@ private ConcatImmutableArray<CodeRefactoringProvider> GetProviders(Document docu
             using (Logger.LogBlock(FunctionId.Refactoring_CodeRefactoringService_GetRefactoringsAsync, cancellationToken))
             {
                 var extensionManager = document.Project.Solution.Workspace.Services.GetRequiredService<IExtensionManager>();
-                var tasks = new List<Task<CodeRefactoring?>>();
+                using var _ = ArrayBuilder<Task<CodeRefactoring?>>.GetInstance(out var tasks);
 
                 foreach (var provider in GetProviders(document))
```
Comment (sharwell) [inline src/Features/Core/Portable/CodeRefactorings/CodeRefactoringService.cs:130]: 📝 `GetProviders` returns a collection with a known length. A fixed-size `Task<CodeRefactoring?>` array would be more efficient here.

**Comment by sharwell** [comment]
I prefer to not make such sweeping changes without one of the following:

1. Direct evidence that the change is needed (i.e. a profiling scenario that demonstrates the specific allocation in question as a meaningful proportion of the scenario total)
2. Analyzer validation of correct lifetime usage of the pooled object(s)

Without the above, the approach seems too risky as a general direction to pursue, considering that the use of an object after it has been returned to a shared pool is a well-known source of bugs in unmanaged code that are _extremely_ difficult to track down. Even if the chance of a specific regression is low, the cost of finding a fixing the specific cause is not.

**Comment by CyrusNajmabadi** [comment]
> Even if the chance of a specific regression is low, the cost of finding a fixing the specific cause is not.

We have to be able to judge relative costs of different approaches with pros and cons.  We've literally had 0 cases of what you're describing with use-after-free, where we've had many issues with using non-pooled collections. 

I hear your concern, but we should be able to make reasonable consistency choices cheaply instead of having every step be so arduous and costly.  History has shown us that following the latter approach just leads to ossification of teh codebase, which works out to be much worse for the overall health versus any minor regressions that appear with bulk moving things forward.  

Right now our codebase if full of debts like this.  Changes are made piecemeal and addressing these inconsistencies constantly faces pushback that leads to nothing happening, and the debt mounting.

Your feedback is def noted.  But balanced against the actual experience i feel we have here, it's overwrought.

**Comment by jasonmalinowski** [comment]
> We've literally had 0 cases of what you're describing with use-after-free, where we've had many issues with using non-pooled collections.

Oh we've totally shipped bugs due to that: https://github.com/dotnet/roslyn/pull/39729

**Comment by CyrusNajmabadi** [comment]
> Oh we've totally shipped bugs due to that: #39729

By 'we', i meant IDE.

And, regardless, each approach has pros and cons.  I've fixed *far more* memory issues around non-pooled collections versus the zero issues around use-after-free.  If we had a perfect approach for all situations, i would use it.  But we don't, so i'm trying ot make the codebase consistent with the form that seems to work better overall and has far less problems

**Comment by sharwell** [comment]
#43255 was caused by an error in this pool. I was not able to determine the root cause.

**Comment by CyrusNajmabadi** [comment]
> in this pool

Which pool?

**Comment by CyrusNajmabadi** [comment]
@jasonmalinowski can i get eyes?

File: src/Features/Core/Portable/AddImport/SymbolReferenceFinder.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/AddImport/SymbolReferenceFinder.cs:122]: i'm ok just keeping the logic simple here :)

File: src/Features/Core/Portable/AddParameter/AddParameterService.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/AddParameter/AddParameterService.cs:276]: i'm ok just keeping the logic here.  trying to contort the code to handle those two cases specially doesn't seem valuble.

File: src/Features/Core/Portable/CodeRefactorings/CodeRefactoringService.cs (hunk 3 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/CodeRefactorings/CodeRefactoringService.cs:130]: that would result in an array that then needs to be gc'ed.  i'd prefer to we in general pool our temp storage so we can just keep reusing these same smallish temp buffers for all successive calls.

File: src/Features/Core/Portable/GenerateMember/GenerateVariable/AbstractGenerateVariableService.State.cs [hunk 4]
```diff
@@ -414,11 +414,10 @@ private static int GetStatementIndex(ChildSyntaxList children, SyntaxNode statem
                 var enclosingMethodSymbol = semanticDocument.SemanticModel.GetEnclosingSymbol<IMethodSymbol>(SimpleNameOrMemberAccessExpressionOpt.SpanStart, cancellationToken);
                 if (enclosingMethodSymbol != null && enclosingMethodSymbol.TypeParameters != null && enclosingMethodSymbol.TypeParameters.Length != 0)
                 {
-                    var combinedTypeParameters = new List<ITypeParameterSymbol>();
+                    using var _ = ArrayBuilder<ITypeParameterSymbol>.GetInstance(out var combinedTypeParameters);
                     combinedTypeParameters.AddRange(availableTypeParameters);
                     combinedTypeParameters.AddRange(enclosingMethodSymbol.TypeParameters);
-                    LocalType = inferredType.RemoveUnavailableTypeParameters(
-                    compilation, combinedTypeParameters);
```
Comment (jasonmalinowski) [inline src/Features/Core/Portable/GenerateMember/GenerateVariable/AbstractGenerateVariableService.State.cs:421]: Oh goodness thank you.

**Comment by jmarolf** [comment]
I am generally in favor of this change. Could we write and analyzer that makes this the preferred style? Any deviation from this style should be accompanied by a command explaining it anyway so adding a pragma disable along with the explanation seems fine to me

**Comment by mavasani** [comment]
@jmarolf We can take couple of approaches:

1. Use the [Banned API analyzer](https://github.com/dotnet/roslyn-analyzers/blob/master/src/Microsoft.CodeAnalysis.BannedApiAnalyzers/BannedApiAnalyzers.Help.md) : already enabled for the repo - so would just need to add a `BannedSymbols.txt` for relevant projects with the `List` .ctor marked as banned. For example, see https://github.com/dotnet/roslyn/blob/master/src/CodeStyle/BannedSymbols.txt

2. Write a new, disabled by default analyzer at https://github.com/dotnet/roslyn-analyzers/tree/master/src/Roslyn.Diagnostics.Analyzers and enable it for IDE layer projects where we want.

**Comment by CyrusNajmabadi** [comment]
Yeah, i'll work on an analyzer when i get a chance :)

**Pull request merged**